mkosi/distributions/centos.py

```diff
@@ -169,43 +218,46 @@ def repository_variants(cls, context: Context, repo: str) -> Iterable[RpmReposit
                 yield RpmRepository(
                     repo.lower(),
                     f"{url}?arch=$basearch&repo=centos-extras-sig-extras-common-$stream",
-                    cls.gpgurls(context),
+                    gpgurls,
                 )
                 yield RpmRepository(
                     f"{repo.lower()}-source",
                     f"{url}?arch=source&repo=centos-extras-sig-extras-common-source-$stream",
-                    cls.gpgurls(context),
+                    gpgurls,
                     enabled=False,
                 )
             else:
                 yield RpmRepository(
                     repo.lower(),
                     f"{url}?arch=$basearch&repo=centos-{repo.lower()}-$stream",
-                    cls.gpgurls(context),
+                    gpgurls,
                 )
                 yield RpmRepository(
                     f"{repo.lower()}-debuginfo",
                     f"{url}?arch=$basearch&repo=centos-{repo.lower()}-debug-$stream",
-                    cls.gpgurls(context),
+                    gpgurls,
                     enabled=False,
                 )
                 yield RpmRepository(
                     f"{repo.lower()}-source",
                     f"{url}?arch=source&repo=centos-{repo.lower()}-source-$stream",
-                    cls.gpgurls(context),
+                    gpgurls,
                     enabled=False,
                 )
 
     @classmethod
     def repositories(cls, context: Context) -> Iterable[RpmRepository]:
+        gpgurls = cls.gpgurls(context)
```

Comment:
Instead of passing them around, wouldn't it work to decorate the method with `functool.lru_cache`? This could be a much smaller change.

Reply:
I don't think we can, requires inputs to be properly cacheable.

Reply:
True, but that's something that could be achieved, I think. All inputs to `Context` are hashable (Args and Config thanks to being frozen), the only thing that makes `Context` not hashable, should then be the lowerdir list, but a hash function like this might just work:
```python
def __hash__(self) -> int:
     return hash(
         (
             self.args,
             self.config,
             self.workspace,
             self.resources,
             self.keyring_dir,
             self.metadata_dir,
             self.package_dir,
             tuple(self.lowerdirs),
             self.upperdir,
             self.workdir,
         ) 
```         
since then all the ingredients are hashable.